server/app/services/rpc_server.rb

```diff
@@ -36,68 +68,98 @@ def initialize(code, message)
 
   # @param [SizedQueue] queue
   def initialize(autostart: true)
-    @queue = self.class.queue
     @handlers = {}
-    @counter = 0
-    @processing = false
+    @msg_counter = 0
+
+    info "initialized"
+
+    async.process if autostart
+    async.watch_queue if autostart
+  end
+
+  def queue
+    self.class.queue
+  end
+
+  def read_and_clear_msg_counter
+    @msg_counter.tap do
+      # this is not atomic, but who cares
+      @msg_counter = 0
+    end
+  end
 
-    async.process! if autostart
+  def watch_queue
+    every(QUEUE_WATCH_PERIOD) do
+      queue_size = self.queue.size
+      msg_dropped = self.class.read_and_clear_msg_dropped
+      msg_counter = self.read_and_clear_msg_counter
+
+      warn "#{queue_size} messages in queue" if queue_size  > QUEUE_DROP_NOTIFICATIONS_LIMIT
+      warn "#{msg_dropped} dropped notifications" if msg_dropped > 0
+      info "#{msg_counter / QUEUE_WATCH_PERIOD} messages per second"
+    end
   end
 
-  def process!
+  def process
     @processing = true
-    while @processing && data = @queue.pop
-      @counter += 1
-      size = data.size
-      if size == 2
-        handle_notification(data[0], data[1])
-      elsif size == 3
-        handle_request(data[0], data[1], data[2])
+
+    # separate thread for blocking queue.pop
+    # XXX: all of the handle_* methods run in the defer thread, not the actor thread...
+    defer {
+      while @processing && data = self.queue.pop
```

Comment:
The `@processing` is still needed, because otherwise the actor crashing and restarting would spawn a new `process => defer { ... }` thread while the old one would keep running, I think...